### PR TITLE
Add project profile

### PR DESCRIPTION
**Pull request by stalhaali**
Implemented:
- Unique project profile for all projects
    - Displays project title, location, category, member capacity, description and project owner
    - List of current members in the project.
- Invited users can join project from project profile
- API route `GET /api/v1/projects/:id` for fetching project.
- API route `POST /api/v1/projects/:pid/join/ for lettings users join the project.

### CODE REVIEW

File: Buddies.API/Controllers/ProjectsController.cs [hunk 1]
```diff
@@ -243,7 +245,209 @@ public async Task<ActionResult<SearchResponse>> GetUser(string search, int page,
...
+            var owner = await _context.Profiles.FindAsync(project.OwnerId);
+            if (owner == null)
+            {
+                return NotFound("OWNER OF PROJECT PROFILE NOT FOUND");
+            }
+            var ownerUser = await _context.Users.FindAsync(project.OwnerId);
+            if (ownerUser == null)
+            {
+                return NotFound("OWNER OF PROJECT PROFILE NOT FOUND");
+            }
```
Comment (miosenpai) [inline Buddies.API/Controllers/ProjectsController.cs]: I'm not sure if this is really the best implementation. If you define the owner as a Profile/User object in Project (following [EF docs](https://docs.microsoft.com/en-us/ef/core/modeling/relationships?tabs=fluent-api%2Cfluent-api-simple-key%2Csimple-key) + [loading it properly](https://docs.microsoft.com/en-us/ef/core/querying/related-data/)) then require an owner to be passed in Project's constructor we can avoid these cases.

File: Buddies.API/Controllers/ProjectsController.cs [hunk 2]
```diff
@@ -243,7 +245,209 @@ public async Task<ActionResult<SearchResponse>> GetUser(string search, int page,
...
+                Title = project.Title,
+                Description = project.Description,
+                Location = project.Location,
+                Username = owner.FirstName + " " + owner.LastName,
+                Email = ownerUser.Email,
+                Category = project.Category,
+                MaxMembers = project.MaxMembers,
+            };
+
+            var membersInProject = _context.Projects.Where(p => p.ProjectId == id).SelectMany(p => p.Members).ToList();
```
Comment (miosenpai) [inline Buddies.API/Controllers/ProjectsController.cs]: Doesn't Project have a [list of members](https://github.com/GDSCUTM-CommunityProjects/Buddies/blob/048063f57e17a951c99dad49bb836e924f7bbd2d/Buddies.API/Entities/Project.cs#L62), do you need to query it again?

File: Buddies.API/Controllers/ProjectsController.cs [hunk 3]
```diff
@@ -243,7 +245,209 @@ public async Task<ActionResult<SearchResponse>> GetUser(string search, int page,
...
+                }
+
+                userInfo.FirstName = userprofile.FirstName;
+                userInfo.LastName = userprofile.LastName;
+                userInfo.Email = member.Email;
+                userInfo.UserId = member.Id;
+                profileResponse.Members.Add(userInfo);
+            }
+
+            var invitedUsers = _context.Projects.Where(p => p.ProjectId == id).SelectMany(p => p.InvitedUsers).ToList();
```
Comment (miosenpai) [inline Buddies.API/Controllers/ProjectsController.cs]: Same as https://github.com/GDSCUTM-CommunityProjects/Buddies/pull/31/files#r825387450.

File: Buddies.API/Controllers/ProjectsController.cs [hunk 4]
```diff
@@ -243,7 +245,209 @@ public async Task<ActionResult<SearchResponse>> GetUser(string search, int page,
...
+            {
+                return NotFound("Invalid Location");
+            }
+
+
+            var requestCategory = await _context.Categories.FirstOrDefaultAsync(p => p.Name == profile.Category);
+            if (requestCategory == null)
+            {
+                return NotFound("Invalid Category");
+            }
```
Comment (miosenpai) [inline Buddies.API/Controllers/ProjectsController.cs]: Same as https://github.com/GDSCUTM-CommunityProjects/Buddies/pull/31/files#r825387298.

File: Buddies.API/Controllers/ProjectsController.cs [hunk 5]
```diff
@@ -243,7 +245,209 @@ public async Task<ActionResult<SearchResponse>> GetUser(string search, int page,
...
+        public async Task<ActionResult> DeleteProjectMember(int pid, int uid)
+        {
+            var project = await _context.Projects.FindAsync(pid);
+
+            if (project == null)
+            {
+                return NotFound("PROFILE NOT FOUND");
+            }
+            if (project.OwnerId != _userManager.GetUserAsync(User).Result.Id) { return Unauthorized(); }
+            var membersInProject = _context.Projects.Where(p => p.ProjectId == pid).SelectMany(p => p.Members).ToList();
```
Comment (miosenpai) [inline Buddies.API/Controllers/ProjectsController.cs]: Same as https://github.com/GDSCUTM-CommunityProjects/Buddies/pull/31/files#r825387298.

File: Buddies.API/Controllers/ProjectsController.cs [hunk 6]
```diff
@@ -243,7 +245,209 @@ public async Task<ActionResult<SearchResponse>> GetUser(string search, int page,
...
+        {
+            var project = await _context.Projects.FindAsync(pid);
+
+            if (project == null)
+            {
+                return NotFound("PROFILE NOT FOUND");
+            }
+            if (project.OwnerId != _userManager.GetUserAsync(User).Result.Id) { return Unauthorized(); }
+
+            var invitedUsers = _context.Projects.Where(p => p.ProjectId == pid).SelectMany(p => p.InvitedUsers).ToList();
```
Comment (miosenpai) [inline Buddies.API/Controllers/ProjectsController.cs]: Same as https://github.com/GDSCUTM-CommunityProjects/Buddies/pull/31/files#r825387298.

File: Buddies.API/Controllers/ProjectsController.cs [hunk 7]
```diff
@@ -243,7 +245,209 @@ public async Task<ActionResult<SearchResponse>> GetUser(string search, int page,
...
+            var project = await _context.Projects.FindAsync(pid);
+
+            if (project == null)
+            {
+                return NotFound("PROFILE NOT FOUND");
+            }
+            if (project.OwnerId == _userManager.GetUserAsync(User).Result.Id) { return BadRequest("Cannot remove owner from project"); }
+
+            var invitedUsers = _context.Projects.Where(p => p.ProjectId == pid).SelectMany(p => p.InvitedUsers).ToList();
+            var membersInProject = _context.Projects.Where(p => p.ProjectId == pid).SelectMany(p => p.Members).ToList();
```
Comment (miosenpai) [inline Buddies.API/Controllers/ProjectsController.cs]: Same as https://github.com/GDSCUTM-CommunityProjects/Buddies/pull/31/files#r825387298.

File: Buddies.API/IO/ProjectProfileResponse.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        [Required(ErrorMessage = "A username is required.")]
+        public string Username { get; init; } = default!;
+
+        [Required(ErrorMessage = "An email is required.")]
+        public string Email { get; init; } = default!;
+
+        [Required(ErrorMessage = "A category is required.")]
+        public string Category { get; set; } = default!;
+
+        [Required] // todo: decide on a reasonable max
```
Comment (miosenpai) [inline Buddies.API/IO/ProjectProfileResponse.cs]: Remove invalid todo.

File: Buddies.API/IO/UpdateProjectProfileRequest.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    [Required(ErrorMessage = "A location is required.")]
+    public string Location { get; init; } = default!;
+
+    [Required(ErrorMessage = "A description is required.")]
+    public string Description { get; init; } = default!;
+
+    [Required(ErrorMessage = "A category is required.")]
+    public string Category { get; init; } = default!;
+
+    [Required] // todo: decide on a reasonable max
```
Comment (miosenpai) [inline Buddies.API/IO/UpdateProjectProfileRequest.cs]: Complete this todo.

File: Buddies.Web/src/pages/projects/[pid].tsx [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+
+  /* Gets project by id and then creates necessary global data structures.
+     (memberLst, invitedLst, project object)
+  */
+  function getProject() {
+    if (!(typeof projectId === 'string')) {
+      alert('error');
+      return;
+    }
+    axios.get(baseURL + projectId).then((res) => {
```
Comment (miosenpai) [inline Buddies.Web/src/pages/projects/[pid].tsx]: Move the API calling logic to `api/index.ts` for consistency and reusability.

File: Buddies.Web/src/pages/projects/[pid].tsx [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+      alert(error);
+    });
+  }
+
+  const addMember = async () => {
+    if (!(typeof projectId === 'string')) {
+      alert('error');
+      return;
+    }
+    const res = await axios.post(`${baseURL + projectId}/join/`, currId).catch((error) => {
```
Comment (miosenpai) [inline Buddies.Web/src/pages/projects/[pid].tsx]: Same as https://github.com/GDSCUTM-CommunityProjects/Buddies/pull/31/files#r825388202.

File: Buddies.Web/src/pages/projects/[pid].tsx [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+      alert(error);
+    });
+    if (true && res) {
```
Comment (miosenpai) [inline Buddies.Web/src/pages/projects/[pid].tsx]: Does res by itself not work?

File: Buddies.Web/src/pages/projects/[pid].tsx (hunk 10 above)
Comment (stalhaali) [reply Buddies.Web/src/pages/projects/[pid].tsx]: Fixed

File: Buddies.Web/src/pages/projects/[pid].tsx (hunk 11 above)
Comment (stalhaali) [reply Buddies.Web/src/pages/projects/[pid].tsx]: Fixed

File: Buddies.Web/src/pages/projects/[pid].tsx (hunk 12 above)
Comment (stalhaali) [reply Buddies.Web/src/pages/projects/[pid].tsx]: It does, I fixed it now. I have no clue why I had "true && res". I think, I had some typescript error and using "true && res" made it go away. My vs code might've just been lagging.

File: Buddies.API/Controllers/ProjectsController.cs (hunk 1 above)
Comment (TianYueDong) [reply Buddies.API/Controllers/ProjectsController.cs]: this is a great idea!

**Review by TianYueDong — APPROVED** [review]
looks good!

File: Buddies.Web/src/components/ProjectSidebar.tsx [hunk 13]
```diff
@@ -0,0 +1,82 @@
+import { ListItemIcon, ListItemText, Typography } from '@mui/material';
+import Container from '@material-ui/core/Container';
+import Button from '@mui/material/Button';
+import Drawer from '@mui/material/Drawer';
+import List from '@mui/material/List';
+import ListItem from '@mui/material/ListItem';
+import Avatar from '@mui/material/Avatar';
+import React from 'react';
```
Comment (andy9775) [inline Buddies.Web/src/components/ProjectSidebar.tsx]: Nit: This should be the first import.

File: Buddies.Web/src/components/ProjectSidebar.tsx [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+  ];
+
+  const list = () => (
+    <div>
+      <Container>
+        <Typography variant="h5" align="center" sx={{ marginTop: 2 }}>Manage Project</Typography>
+      </Container>
+      <Container sx={{ marginTop: 4 }}>
+        <Avatar />
+        <Typography sx={{ marginTop: 1 }}>John Doe</Typography>
```
Comment (andy9775) [inline Buddies.Web/src/components/ProjectSidebar.tsx]: Forgot to change the placeholder text?

File: Buddies.Web/src/pages/projects/[pid].tsx [hunk 15]
```diff
@@ -0,0 +1,235 @@
+import Grid from '@material-ui/core/grid';
+import Container from '@material-ui/core/Container';
+import Card from '@mui/material/Card';
+import CardActionArea from '@mui/material/CardActionArea';
+import Typography from '@mui/material/Typography';
+import Avatar from '@mui/material/Avatar';
+import AddIcon from '@mui/icons-material/Add';
+import EmailIcon from '@mui/icons-material/Email';
+import React, { useEffect } from 'react';
```
Comment (andy9775) [inline Buddies.Web/src/pages/projects/[pid].tsx]: ditto

**Review by andy9775 — APPROVED** [review]
Just some minor nits. Otherwise LGTM!

**Pull request merged**